### PR TITLE
tests: add run-openshift template tests

### PR DESCRIPTION
**Pull request by pkubatrh**
Resolves: #204

### CODE REVIEW

**Comment by pkubatrh** [comment]
@omron93 Are the tests run when they are present or do we have to trigger them manually in some way?

File: test/run-openshift [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+#
+
+THISDIR=$(dirname ${BASH_SOURCE[0]})
+
+source ${THISDIR}/test-lib-openshift.sh
+
+set -exo nounset
+
+function check_postgresql_os_service_connection() {
+  local util_image_name="${1}" ; shift
```
Comment (praiskup) [inline test/run-openshift]: Nit:please use just "foo=$baz" syntax.  Looks much nicer, the quotes and brackets are not needed in this case.

File: test/run-openshift [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+#
+# Test the Postgresql image in OpenShift.
+#
+# IMAGE_NAME specifies a name of the candidate image used for testing.
+# The image has to be available before this script is executed.
+#
+
+THISDIR=$(dirname ${BASH_SOURCE[0]})
+
+source ${THISDIR}/test-lib-openshift.sh
```
Comment (praiskup) [inline test/run-openshift]: I would rather quote $THISDIR.

File: test/run-openshift [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  local user="${1}" ; shift
+  local pass="${1}" ; shift
+  local database="${1}" ; shift
+  local timeout="${1:-60}" ; shift || :
+  local pod_ip=$(ct_os_get_service_ip ${service_name})
+
+  : "  Service ${service_name} check ..."
+
+  local cmd="echo 'SELECT 42 as testval' | PGPASSWORD=${pass} PGCONNECT_TIMEOUT=15 \
+             psql -t -h ${pod_ip} -U ${user} -d ${database}"
```
Comment (praiskup) [inline test/run-openshift]: Could we use `pg_isready`, since that will be the live check in future, too?

File: test/run-openshift [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    echo -n "."
+    [ ${SECONDS} -gt ${timeout} ] && break
+    sleep 3
+  done
+  echo " FAIL"
+  return 1
+}
+
+function test_postgresql_pure_image() {
+  local image_name=${1:-centos/postgresql-96-centos7}
```
Comment (praiskup) [inline test/run-openshift]: The default value seems to be unused, so `image_name=$1` would look a bit safer.

File: test/run-openshift [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  ct_os_deploy_template_image ${THISDIR}/postgresql-ephemeral-template.json \
+                              DATABASE_SERVICE_NAME="${service_name}" \
+                              POSTGRESQL_USER=testu \
+                              POSTGRESQL_PASSWORD=testp \
+                              POSTGRESQL_DATABASE=testdb
+
+  ct_os_wait_pod_ready "${service_name}" 60
+  check_postgresql_os_service_connection "${image_name}" "${service_name}" testu testp testdb
+
+  ct_os_delete_project
```
Comment (praiskup) [inline test/run-openshift]: Nice, this seems to be promising!  I'm looking forward to see the library code.

Comment (praiskup) [reply test/run-openshift]: Ah, it's already there :-) my bad.

**Comment by praiskup** [comment]
Seems like the 'test-openshift' target used to be in https://github.com/sclorg/rhscl-container-ci/blob/master/yaml/jobs/collections/mongodb-rh.yaml so the openshift tests used to be triggered automatically for mongo.  Better approach could be to invoke the `tests-openshift` target for every image, and that target could succeed with "not-implemented" message if the image is not yet CI-tested for openshift.

**Comment by pkubatrh** [comment]
@praiskup Made some changes according to your suggestions, thanks!

**Comment by pkubatrh** [comment]
Seems like the code works on both rhel and centos-based Openshift installations. Only the library does not yet provide any functions to install openshift on a rhel7 system.

**Comment by pkubatrh** [comment]
@omron93 Since `make test-openshift` should be a no-op if the openshift tests are not present can we enable the make target to be triggered by default on every image in CI?

**Comment by omron93** [comment]
> @omron93 Since make test-openshift should be a no-op if the openshift tests are not present can we enable the make target to be triggered by default on every image in CI?

OK. I've enabled it for all images. Test **has to be explicitly triggered** by [test-openshift]

> Seems like the code works on both rhel and centos-based Openshift installations. Only the library does not yet provide any functions to install openshift on a rhel7 system.

I would let installing oc tool to the user (and to the CI), same we do with `docker-squash`. It simplify our tests and avoid problems with different environments.

**Comment by pkubatrh** [comment]
>RHEL CI is failing because you try to install centos package (CI image already contain oc).

Looking at the logs the centos image does not seem to have openshift installed by default. Are we going to update it so it does or should I just make the install function call specific to centos?

**Comment by pkubatrh** [comment]
>In CentOS there is problem with starting cluster. I'll investigate it tomorrow.

The same issue is in rhel, it happens only the first time the `cluster up` action is called. No idea how to work around it other than just not failing the first time and trying again...

**Comment by omron93** [comment]
> The same issue is in rhel, it happens only the first time the cluster up action is called. No idea how to work around it other than just not failing the first time and trying again...

iptables has to be flushed. Please take a look at https://github.com/sclorg/container-common-scripts/pull/40

With this it should work. No `ct_os_install_in_centos` is needed (I would let this to user to setup testing machine - number of OS version,... is too big). So only `ct_os_cluster_up` is enought to start OpenShift in CI.

**Comment by pkubatrh** [comment]
[test-openshift]

**Comment by pkubatrh** [comment]
Seems like the openshift test fails now on `docker login`

**Comment by pkubatrh** [comment]
[test-openshift]

**Comment by omron93** [comment]
> Seems like the openshift test fails now on docker login

It seems that it takes some time. In last centos openshift tests it succeeded in fifth attempt.

I'll try to find out what is causing this.

**Comment by pkubatrh** [comment]
[test-openshift]

**Comment by pkubatrh** [comment]
[test-openshift]

**Comment by pkubatrh** [comment]
@praiskup @omron93 For some reason I do not see rhel7 CI starting.

**Comment by omron93** [comment]
> For some reason I do not see rhel7 CI starting.

There was Jenkins upgrade by Central CI admins. But problems with plugins occurred so we have to wait.

**Comment by pkubatrh** [comment]
All checks passing now

**Comment by praiskup** [comment]
Looks fine.

**Comment by pkubatrh** [comment]
Thanks for review! Merging

File: test/run-openshift [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+                          --env POSTGRESQL_ADMIN_PASSWORD=test
+
+  ct_os_wait_pod_ready "${service_name}" 60
+  check_postgresql_os_service_connection "${image_name}" "${service_name}" postgres test postgres
+
+  ct_os_delete_project
+}
+
+function test_postgresql_template() {
+  local image_name=${1:-centos/postgresql-96-centos7}
```
Comment (pvalena) [inline test/run-openshift]: Why is there 'backup' image_name?
IMHO this should be tested specifically with the CI-built image.

**Pull request merged**